### ISSUE TITLE
Floating point errors with devicePixelRatio

### ISSUE THREAD

**Issue by ds84182** [flutter/flutter#12538]
The Display Size feature in Android N allows users to choose a screen DPI multiplier from 0.85 to 1.50. My device, which has a pixel ratio of 3.0. With a multiplier of 0.85 my pixel ratio drops to 2.55... however, Flutter reports a pixel ratio of 2.549999952316284.

I can reproduce this floating point error using the following Dart code:

```dart
print((new Float32List(1)..[0] = 2.55)[0]);
```

The pixel ratio is (probably) being casted to a float somewhere.

This also gives me white or black lines between the colors in Flutter Gallery's Colors demo. When the pixel ratio is 2.5 or 3.0 the lines cannot be seen (although I've been able to get the lines to appear _sometimes_ if I do a short fling... maybe scroll offsets aren't being pixel corrected?).

[Screenshot with my pixel ratio at 2.55](https://i.imgur.com/j2EOUNO.png)

[Screenshot with my pixel ratio at 3.0](https://i.imgur.com/XkUVhfS.png)

**Comment by Hixie** [comment]
"2.55" isn't representable in binary. The closest number you can get in IEEE-754 64bit floating point is 2.5499999523162841796875. So that's what we have to use.

What you're seeing in those screenshots is what happens when we have to do subpixel antialiasing because the boxes don't quite line up with the physical pixels. We end up antialiasing with the color underneath (which in this case is lighter, so you see lighter pixels). I'm not sure what we could really do about this. If we snapped every box to fit the physical pixels you'd avoid this problem but get much worse other problems (like jittering, uneven sizes, etc).

In conclusion, this is all working as intended.

**Comment by zoechi** [comment]
@Hixie Should this issue be closed then?

**Comment by Hixie** [comment]
I guess so. @ds84182 please let us know if there's anything unclear in the comments above or if we can help in any way.
